fish-rust/src/fish.rs

```diff
@@ -0,0 +1,877 @@
+                    base.pop();
+                    base
+                };
+
+                paths = if seems_installed {
+                    ConfigPaths {
+                        data: base_path.join("share/fish"),
+                        sysconf: base_path.join("etc"),
+                        doc: base_path.join("share/doc/fish"),
+                        bin: base_path.join("bin"),
+                    }
+                } else {
+                    ConfigPaths {
+                        data: base_path.join("share"),
+                        sysconf: base_path.join("etc"),
+                        doc: base_path.join("user_doc/html"),
+                        bin: base_path,
+                    }
+                };
+
+                if paths.data.exists() && paths.sysconf.exists() {
+                    // The docs dir may not exist; in that case fall back to the compiled in path.
+                    if !paths.doc.exists() {
+                        paths.doc = PathBuf::from(DOC_DIR);
+                    }
+                    done = true;
+                }
+            }
+        }
+    }
+
+    if !done {
+        // Fall back to what got compiled in.
+        FLOG!(config, "Using compiled in paths:");
+        paths = ConfigPaths {
+            data: PathBuf::from(DATA_DIR).join("fish"),
+            sysconf: PathBuf::from(SYSCONF_DIR).join("fish"),
+            doc: DOC_DIR.into(),
+            bin: BIN_DIR.into(),
+        }
+    }
+
+    FLOGF!(
+        config,
+        sprintf!(
+            "determine_config_directory_paths() results:\npaths.data: %ls\npaths.sysconf: \
+        %ls\npaths.doc: %ls\npaths.bin: %ls",
+            paths.data.display().to_string(),
+            paths.sysconf.display().to_string(),
+            paths.doc.display().to_string(),
+            paths.bin.display().to_string()
+        )
+    );
+
+    paths
+}
+
+// Source the file config.fish in the given directory.
+fn source_config_in_directory(parser: &mut ffi::parser_t, dir: &wstr) {
+    // If the config.fish file doesn't exist or isn't readable silently return. Fish versions up
+    // thru 2.2.0 would instead try to source the file with stderr redirected to /dev/null to deal
+    // with that possibility.
+    //
+    // This introduces a race condition since the readability of the file can change between this
+    // test and the execution of the 'source' command. However, that is not a security problem in
+    // this context so we ignore it.
+    let config_pathname = dir.to_owned() + L!("/config.fish");
+    let escaped_pathname = escape_string(dir, EscapeStringStyle::default()) + L!("/config.fish");
+    if waccess(&config_pathname, libc::R_OK) != 0 {
+        FLOGF!(
+            config,
+            "not sourcing %ls (not readable or does not exist)",
+            escaped_pathname
+        );
+        return;
+    }
+    FLOGF!(config, "sourcing", escaped_pathname);
```

Comment:
Bad format string here.

Reply:
Should probably just be `FLOG!` - which joins the arguments.